### ISSUE TITLE
ONNX and PT, inconsistent results

### ISSUE THREAD

**Issue by monkeycc** [ultralytics/ultralytics#8666]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

ONNX and PT, inconsistent results

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-OpenCV-ONNX-Python
`CLASSES = ["aa","bb"]`
`python main.py --model yolov8n.onnx --img image.jpg
`

```python

from ultralytics import YOLO

# Load a model
model = YOLO('yolov8s.pt')

results = model.train('xxx.yaml',imgsz=640)

results = model.val()

success = model.export(format="onnx")

```

![d](https://github.com/ultralytics/ultralytics/assets/6490927/cf3629e1-ea90-4b56-9888-189048185197)

**Comment by monkeycc** [comment]
Unfortunately, the problem still exists

`onnxsim xx.onnx xx_simplified.onnx`

```

┏━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃            ┃ Original Model ┃ Simplified Model ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│ Add        │ 9              │ 8                │
│ Concat     │ 19             │ 19               │
│ Constant   │ 149            │ 143              │
│ Conv       │ 64             │ 64               │
│ Div        │ 2              │ 1                │
│ Gather     │ 1              │ 0                │
│ MaxPool    │ 3              │ 3                │
│ Mul        │ 60             │ 58               │
│ Reshape    │ 5              │ 5                │
│ Resize     │ 2              │ 2                │
│ Shape      │ 1              │ 0                │
│ Sigmoid    │ 58             │ 58               │
│ Slice      │ 2              │ 2                │
│ Softmax    │ 1              │ 1                │
│ Split      │ 9              │ 9                │
│ Sub        │ 2              │ 2                │
│ Transpose  │ 1              │ 1                │
│ Model Size │ 42.7MiB        │ 42.6MiB          │
└────────────┴────────────────┴──────────────────┘
```

---------------

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-OpenCV-ONNX-Python
I'm not sure how to modify the code


https://mm-1301360300.cos.ap-guangzhou.myqcloud.com/onnx_pt.zip
I suggest downloading the model file to reproduce this issue

@glenn-jocher

**Comment by glenn-jocher** [comment]
@monkeycc hey there! Thanks for giving the simplification a shot and sharing the details. 🌟 It looks like the simplification did tweak the model a bit, but if the inconsistency persists, it might be rooted deeper than just the model's structure.

For the code modification part, especially when working with the YOLOv8 OpenCV ONNX Python example, ensure you're adjusting the input preprocessing and output postprocessing to match exactly what your model expects and outputs. This often involves ensuring the image normalization and scaling are consistent with how the model was trained.

Regarding the model file, I'll definitely take a look to better understand the issue. In the meantime, could you double-check the preprocessing steps? Here's a quick snippet for reference:

```python
import cv2
import numpy as np

# Assuming 'frame' is your input image
frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
frame = cv2.resize(frame, (640, 640))  # Resize as per model's expected input
frame = frame / 255.0  # Normalize if your model expects pixel values in [0,1]
frame = np.transpose(frame, (2, 0, 1))  # Change data layout from HWC to CHW
frame = np.expand_dims(frame, axis=0)  # Add batch dimension

# Now 'frame' is ready to be passed to the model
```

Make sure the output processing aligns with how the model's predictions are structured. This often involves interpreting the raw output tensors correctly.

I'll dive into the model files you've shared to get a better grasp. Thanks for your patience and for providing detailed info! 🛠️

**Comment by monkeycc** [comment]
the non-max
suppression algorithm does not check the box classes and therefore it
ignores the inner box

**Comment by glenn-jocher** [comment]
@monkeycc hey there! 👋 It sounds like you're encountering an issue where non-max suppression (NMS) isn't considering the classes of the bounding boxes, leading to it ignoring inner boxes of different classes. 

NMS should indeed consider classes to ensure it suppresses overlapping boxes of the same class while allowing boxes of different classes to coexist. Here's a quick snippet to ensure class-wise NMS:

```python
import torch

def class_wise_nms(boxes, scores, iou_threshold, class_ids):
    unique_classes = class_ids.unique()
    keep_boxes = []

    for cls in unique_classes:
        # Filter boxes and scores by class
        cls_mask = class_ids == cls
        cls_boxes = boxes[cls_mask]
        cls_scores = scores[cls_mask]

        # Perform NMS
        keep = torch.ops.torchvision.nms(cls_boxes, cls_scores, iou_threshold)

        keep_boxes.append(keep)

    # Concatenate kept indices from all classes
    keep_boxes = torch.cat(keep_boxes)
    return keep_boxes

# Example usage
boxes = torch.tensor([[10, 10, 50, 50], [20, 20, 60, 60], [30, 30, 70, 70]])
scores = torch.tensor([0.9, 0.8, 0.7])
class_ids = torch.tensor([0, 1, 0])
iou_threshold = 0.5

keep = class_wise_nms(boxes, scores, iou_threshold, class_ids)
```

This ensures that NMS is performed separately for each class, allowing for correct suppression based on class. Hope this helps! Let us know if you have further questions. 😊

**Comment by monkeycc** [comment]
https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-ONNXRuntime

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-OpenCV-ONNX-Python

Need to repair

**Comment by glenn-jocher** [comment]
@monkeycc hey there! 👋 Thanks for pointing out the need for updates in those examples. We're always working to improve and keep our documentation and examples fresh and accurate. Could you specify what issues or errors you're encountering? This will help us target the fixes more effectively. 

In the meantime, here's a quick snippet to get you started with YOLOv8 ONNX models using ONNXRuntime:

```python
import onnxruntime
import numpy as np

# Load the ONNX model
session = onnxruntime.InferenceSession("yolov8.onnx")

# Prepare your input according to the model's requirements
# Example for a single image of shape (1, 3, 640, 640)
input_image = np.random.rand(1, 3, 640, 640).astype(np.float32)

# Run inference
outputs = session.run(None, {"input": input_image})

# Process the outputs as needed
```

And for using YOLOv8 ONNX models with OpenCV:

```python
import cv2
import numpy as np

# Load the ONNX model
net = cv2.dnn.readNetFromONNX("yolov8.onnx")

# Prepare your input according to the model's requirements
# Example for a single image of shape (1, 3, 640, 640)
input_image = np.random.rand(1, 3, 640, 640).astype(np.float32)

# Set the input to the model
net.setInput(input_image)

# Run inference
outputs = net.forward()

# Process the outputs as needed
```

Let us know if there's anything specific you need help with! 😊

**Comment by monkeycc** [comment]
I suspect you are ChatGPT answering ...

**Comment by glenn-jocher** [comment]
@monkeycc hey there! 😊 If you have any questions or need assistance with YOLOv8, feel free to drop them here. Whether it's about model training, inference, or anything else YOLOv8 related, I'm here to help. Let's make object detection awesome together! 🚀
